File: web/src/StreamInformation.cpp
/*
 * Copyright (c) 2022, 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#include "StreamInformation.h"


ffmpegkit::StreamInformation::StreamInformation(
    std::shared_ptr<ffmpegkit::json::Value> streamInformationValue)
    : _streamInformationValue{streamInformationValue} {}

std::shared_ptr<int64_t> ffmpegkit::StreamInformation::getIndex() {
    return getNumberProperty(KeyIndex);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getType() {
    return getStringProperty(KeyType);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getCodec() {
    return getStringProperty(KeyCodec);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getCodecLong() {
    return getStringProperty(KeyCodecLong);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getFormat() {
    return getStringProperty(KeyFormat);
}

std::shared_ptr<int64_t> ffmpegkit::StreamInformation::getWidth() {
    return getNumberProperty(KeyWidth);
}

std::shared_ptr<int64_t> ffmpegkit::StreamInformation::getHeight() {
    return getNumberProperty(KeyHeight);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getBitrate() {
    return getStringProperty(KeyBitRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getSampleRate() {
    return getStringProperty(KeySampleRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getSampleFormat() {
    return getStringProperty(KeySampleFormat);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getChannelLayout() {
    return getStringProperty(KeyChannelLayout);
}

std::shared_ptr<std::string>
ffmpegkit::StreamInformation::getSampleAspectRatio() {
    return getStringProperty(KeySampleAspectRatio);
}

std::shared_ptr<std::string>
ffmpegkit::StreamInformation::getDisplayAspectRatio() {
    return getStringProperty(KeyDisplayAspectRatio);
}

std::shared_ptr<std::string>
ffmpegkit::StreamInformation::getAverageFrameRate() {
    return getStringProperty(KeyAverageFrameRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getRealFrameRate() {
    return getStringProperty(KeyRealFrameRate);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getTimeBase() {
    return getStringProperty(KeyTimeBase);
}

std::shared_ptr<std::string> ffmpegkit::StreamInformation::getCodecTimeBase() {
    return getStringProperty(KeyCodecTimeBase);
}

std::shared_ptr<ffmpegkit::json::Value> ffmpegkit::StreamInformation::getTags() {
    return getProperty(KeyTags);
}

std::shared_ptr<std::string>
ffmpegkit::StreamInformation::getStringProperty(const char *key) {
    if (_streamInformationValue == nullptr) {
        return nullptr;
    }
    const ffmpegkit::json::Value *property = _streamInformationValue->find(key);
    if (property == nullptr) {
        return nullptr;
    }
    return property->getString();
}

std::shared_ptr<int64_t>
ffmpegkit::StreamInformation::getNumberProperty(const char *key) {
    if (_streamInformationValue == nullptr) {
        return nullptr;
    }
    const ffmpegkit::json::Value *property = _streamInformationValue->find(key);
    if (property == nullptr) {
        return nullptr;
    }
    return property->getInt();
}

std::shared_ptr<ffmpegkit::json::Value>
ffmpegkit::StreamInformation::getProperty(const char *key) {
    if (_streamInformationValue == nullptr) {
        return nullptr;
    }
    const ffmpegkit::json::Value *property = _streamInformationValue->find(key);
    if (property == nullptr) {
        return nullptr;
    }
    return std::make_shared<ffmpegkit::json::Value>(*property);
}

std::shared_ptr<ffmpegkit::json::Value>
ffmpegkit::StreamInformation::getAllProperties() {
    if (_streamInformationValue == nullptr) {
        return nullptr;
    }
    return std::make_shared<ffmpegkit::json::Value>(*_streamInformationValue);
}

File: web/js/src/StreamInformation.js
/*
 * Copyright (c) 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

// LEAF MODULE - MUST NOT IMPORT ANYTHING.

export class StreamInformation {
    static KEY_INDEX = 'index';
    static KEY_TYPE = 'codec_type';
    static KEY_CODEC = 'codec_name';
    static KEY_CODEC_LONG = 'codec_long_name';
    static KEY_FORMAT = 'pix_fmt';
    static KEY_WIDTH = 'width';
    static KEY_HEIGHT = 'height';
    static KEY_BIT_RATE = 'bit_rate';
    static KEY_SAMPLE_RATE = 'sample_rate';
    static KEY_SAMPLE_FORMAT = 'sample_fmt';
    static KEY_CHANNEL_LAYOUT = 'channel_layout';
    static KEY_SAMPLE_ASPECT_RATIO = 'sample_aspect_ratio';
    static KEY_DISPLAY_ASPECT_RATIO = 'display_aspect_ratio';
    static KEY_AVERAGE_FRAME_RATE = 'avg_frame_rate';
    static KEY_REAL_FRAME_RATE = 'r_frame_rate';
    static KEY_TIME_BASE = 'time_base';
    static KEY_CODEC_TIME_BASE = 'codec_time_base';
    static KEY_TAGS = 'tags';

    constructor(data) {
        this._allProperties = data ?? null;
    }

    getIndex() {
        return this.getNumberProperty(StreamInformation.KEY_INDEX);
    }

    getType() {
        return this.getStringProperty(StreamInformation.KEY_TYPE);
    }

    getCodec() {
        return this.getStringProperty(StreamInformation.KEY_CODEC);
    }

    getCodecLong() {
        return this.getStringProperty(StreamInformation.KEY_CODEC_LONG);
    }

    getFormat() {
        return this.getStringProperty(StreamInformation.KEY_FORMAT);
    }

    getWidth() {
        return this.getNumberProperty(StreamInformation.KEY_WIDTH);
    }

    getHeight() {
        return this.getNumberProperty(StreamInformation.KEY_HEIGHT);
    }

    getBitrate() {
        return this.getStringProperty(StreamInformation.KEY_BIT_RATE);
    }

    getSampleRate() {
        return this.getStringProperty(StreamInformation.KEY_SAMPLE_RATE);
    }

    getSampleFormat() {
        return this.getStringProperty(StreamInformation.KEY_SAMPLE_FORMAT);
    }

    getChannelLayout() {
        return this.getStringProperty(StreamInformation.KEY_CHANNEL_LAYOUT);
    }

    getSampleAspectRatio() {
        return this.getStringProperty(StreamInformation.KEY_SAMPLE_ASPECT_RATIO);
    }

    getDisplayAspectRatio() {
        return this.getStringProperty(StreamInformation.KEY_DISPLAY_ASPECT_RATIO);
    }

    getAverageFrameRate() {
        return this.getStringProperty(StreamInformation.KEY_AVERAGE_FRAME_RATE);
    }

    getRealFrameRate() {
        return this.getStringProperty(StreamInformation.KEY_REAL_FRAME_RATE);
    }

    getTimeBase() {
        return this.getStringProperty(StreamInformation.KEY_TIME_BASE);
    }

    getCodecTimeBase() {
        return this.getStringProperty(StreamInformation.KEY_CODEC_TIME_BASE);
    }

    getTags() {
        return this.getProperty(StreamInformation.KEY_TAGS);
    }

    getStringProperty(key) {
        return this._getProperty(key);
    }

    getNumberProperty(key) {
        return this._getProperty(key);
    }

    getProperty(key) {
        return this._getProperty(key);
    }

    getAllProperties() {
        return this._allProperties;
    }

    _getProperty(key) {
        if (this._allProperties == null) {
            return null;
        }
        if (Object.prototype.hasOwnProperty.call(this._allProperties, key)) {
            return this._allProperties[key];
        }
        return null;
    }
}
